File: android/src/main/java/com/cloudwebrtc/webrtc/SessionDescription.kt
package com.cloudwebrtc.webrtc

import org.webrtc.SessionDescription as WSessionDescription;

enum class SessionDescriptionType {
    OFFER,
    PRANSWER,
    ANSWER,
    ROLLBACK;

    companion object {
        fun fromWebRtc(type: WSessionDescription.Type): SessionDescriptionType {
            return when (type) {
                WSessionDescription.Type.OFFER -> OFFER
                WSessionDescription.Type.PRANSWER -> PRANSWER
                WSessionDescription.Type.ANSWER -> ANSWER
                WSessionDescription.Type.ROLLBACK -> ROLLBACK
            }
        }
    }

    fun intoWebRtc(): WSessionDescription.Type {
        return when (this) {
            OFFER -> WSessionDescription.Type.OFFER
            PRANSWER -> WSessionDescription.Type.PRANSWER
            ANSWER -> WSessionDescription.Type.ANSWER
            ROLLBACK -> WSessionDescription.Type.ROLLBACK
        }
    }
}

data class SessionDescription(val type: SessionDescriptionType, val description: String) {
    companion object {
        fun fromWebRtc(sdp: WSessionDescription): SessionDescription {
            return SessionDescription(
                    SessionDescriptionType.fromWebRtc(sdp.type),
                    sdp.description
            );
        }
    }

    fun intoWebRtc(): WSessionDescription {
        return WSessionDescription(type.intoWebRtc(), description)
    }
}

File: lib/src/model/session_description.dart
/// Type of the [SessionDescription].
enum SessionDescriptionType {
  /// Indicates that description is the initial proposal in an
  /// offer/answer exchange.
  offer,

  /// Indicates that the description is a provisional answer and may be
  /// changed when the definitive choice will be given.
  pranswer,

  /// Indicates that description is the definitive choice in an
  /// offer/answer exchange.
  answer,

  /// Indicates that description rolls back to offer/answer state to the
  /// last stable state.
  rollback,
}

/// SDP offer which can be set to `PeerConnection`.
class SessionDescription {
  /// Constructs new [SessionDescription] with a provided type and SDP.
  SessionDescription(this.type, this.description);

  /// Creates [SessionDescription] based on the [Map] received from
  /// the native side.
  SessionDescription.fromMap(dynamic map) {
    type = SessionDescriptionType.values[map['type']];
    description = map['description'];
  }

  /// Type of this [SessionDescription].
  late SessionDescriptionType type;

  /// SDP of this [SessionDescription].
  late String description;

  /// Converts this model to the [Map] expected by Flutter.
  dynamic toMap() { // TODO(#31): Map<String, dynamic>?
    return {
      'type': type.index,
      'description': description,
    };
  }
}
